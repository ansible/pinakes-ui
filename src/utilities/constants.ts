// eslint-disable no-undef
export const CATALOG_API_BASE =
  // eslint-disable-next-line no-undef
  DEPLOYMENT_MODE === 'standalone'
    ? // eslint-disable-next-line no-undef
      `${API_HOST}${API_BASE_PATH}`
    : `${process.env.BASE_PATH || '/api'}/catalog/v1.3`;
// eslint-disable-next-line no-undef
export const AUTH_API_BASE =
  // eslint-disable-next-line no-undef
  DEPLOYMENT_MODE === 'standalone'
    ? // eslint-disable-next-line no-undef
      `${API_HOST}${AUTH_BASE_PATH}`
    : `${process.env.BASE_PATH || '/api'}/auth`;

export const SOURCES_API_BASE =
  // eslint-disable-next-line no-undef
  DEPLOYMENT_MODE === 'standalone'
    ? // eslint-disable-next-line no-undef
      `${API_HOST}${API_BASE_PATH}`
    : `${process.env.BASE_PATH || '/api'}/sources/v1.0`;
export const APPROVAL_API_BASE =
  // eslint-disable-next-line no-undef
  DEPLOYMENT_MODE === 'standalone'
    ? // eslint-disable-next-line no-undef
      `${API_HOST}${API_BASE_PATH}`
    : `${process.env.BASE_PATH || '/api'}/approval/v1.2`;
export const CATALOG_INVENTORY_API_BASE =
  // eslint-disable-next-line no-undef
  DEPLOYMENT_MODE === 'standalone'
    ? // eslint-disable-next-line no-undef
      `${API_HOST}${API_BASE_PATH}`
    : `${process.env.BASE_PATH || '/api'}/catalog-inventory/v1.0`;
export const RBAC_API_BASE =
  // eslint-disable-next-line no-undef
  DEPLOYMENT_MODE === 'standalone'
    ? // eslint-disable-next-line no-undef
      `${API_HOST}${API_BASE_PATH}`
    : `${process.env.BASE_PATH || '/api'}/rbac/v1`;

export const APPLICATION_TITLE =
  // eslint-disable-next-line no-undef
  `${process.env.APPLICATION_NAME || APPLICATION_NAME || 'Pinakes'}`;

export const permissionValues = ['delete', 'order', 'read', 'update'];

export const permissionOptions = [
  {
    value: 'delete,order,read,update',
    label: 'Can order/edit'
  },
  {
    value: 'order,read',
    label: 'Can order/view'
  }
];

export const PORTFOLIO_RESOURCE_TYPE = 'Portfolio';
export const PORTFOLIO_ITEM_RESOURCE_TYPE = 'PortfolioItem';
export const INVENTORY_RESOURCE_TYPE = 'ServiceInventory';
export const DEFAULT_MAX_LENGTH = 64;

export const APP_NAME = {
  Portfolio: 'catalog',
  PortfolioItem: 'catalog',
  ServiceInventory: 'inventory'
};

export const BEFORE_TYPE = 'before';
export const AFTER_TYPE = 'after';
export const RETURN_TYPE = 'return';

export const MAX_RETRY_LIMIT = 3;
export const CATALOG_ADMIN_ROLE = 'catalog-admin';
export const APPROVAL_ADMIN_ROLE = 'approval-admin';
export const APPROVAL_APPROVER_ROLE = 'approval-approver';
export const CATALOG_UI_PREFIX = '/ui/catalog';
