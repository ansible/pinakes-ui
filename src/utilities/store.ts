import promiseMiddleware from 'redux-promise-middleware';
import ReducerRegistry, {
  applyReducerHash
} from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';
import { notificationsReducer } from '@redhat-cloud-services/frontend-components-notifications/redux';
import notificationsMiddleware from '@redhat-cloud-services/frontend-components-notifications/notificationsMiddleware';

import reduxLogger from 'redux-logger';
import thunk from 'redux-thunk';
import orderReducer, {
  orderInitialState
} from '../redux/reducers/order-reducer';
import platformReducer, {
  platformInitialState
} from '../redux/reducers/platform-reducer';
import orderProcessReducer, {
  orderProcessInitialState
} from '../redux/reducers/order-process-reducer';
import portfolioReducer, {
  portfoliosInitialState
} from '../redux/reducers/portfolio-reducer';
import approvalReducer, {
  approvalInitialState
} from '../redux/reducers/approval-reducer';
import rbacReducer, { rbacInitialState } from '../redux/reducers/rbac-reducer';
import shareReducer, {
  shareInfoInitialState
} from '../redux/reducers/share-reducer';
import openApiReducer, {
  openApiInitialState
} from '../redux/reducers/open-api-reducer';

import requestReducer, {
  requestsInitialState
  // @ts-ignore
} from '../redux/reducers/request-reducer';
import workflowReducer, {
  workflowsInitialState
  // @ts-ignore
} from '../redux/reducers/workflow-reducer';
import templateReducer, {
  templatesInitialState
  // @ts-ignore
} from '../redux/reducers/template-reducer';
import notificationSettingsReducer, {
  notificationSettingsInitialState
  // @ts-ignore
} from '../redux/reducers/notification-setting-reducer';
import groupReducer, {
  groupsInitialState
  // @ts-ignore
} from '../redux/reducers/group-reducer';
import loadingStateMiddleware from './loading-state-middleware';
import emptyDataMiddleware from './empty-data-middleware';
import breadcrumbsReducer, {
  initialBreadcrumbsState
} from '../redux/reducers/breadcrumbs-reducer';
import i18nReducer, { i18nInitialState } from '../redux/reducers/i18n-reducer';
import viewStateMiddleware from './view-state-middleware';
import unAuthorizedMiddleware from './unauthorized-middleware';
import { Store } from 'redux';

const prodMiddlewares = [
  notificationsMiddleware({
    errorTitleKey: ['errors', 'message', 'statusText'],
    errorDescriptionKey: [
      'data.errors[0].detail',
      'data.errors',
      'data.error',
      'data.message',
      'data.detail',
      'response.body.errors',
      'data',
      'errorMessage',
      'stack',
      'detail'
    ]
  })
];

const baseMiddlewares = [
  thunk,
  promiseMiddleware,
  unAuthorizedMiddleware,
  viewStateMiddleware,
  loadingStateMiddleware,
  emptyDataMiddleware
];
const registerReducers = (registry: ReducerRegistry): void => {
  registry.register({
    orderReducer: applyReducerHash(orderReducer, orderInitialState),
    platformReducer: applyReducerHash(platformReducer, platformInitialState),
    orderProcessReducer: applyReducerHash(
      orderProcessReducer,
      orderProcessInitialState
    ),
    portfolioReducer: applyReducerHash(
      portfolioReducer,
      portfoliosInitialState
    ),
    approvalReducer: applyReducerHash(approvalReducer, approvalInitialState),
    requestReducer: applyReducerHash(requestReducer, requestsInitialState),
    workflowReducer: applyReducerHash(workflowReducer, workflowsInitialState),
    templateReducer: applyReducerHash(templateReducer, templatesInitialState),
    notificationSettingsReducer: applyReducerHash(
      notificationSettingsReducer,
      notificationSettingsInitialState
    ),
    groupReducer: applyReducerHash(groupReducer, groupsInitialState),
    rbacReducer: applyReducerHash(rbacReducer, rbacInitialState),
    shareReducer: applyReducerHash(shareReducer, shareInfoInitialState),
    openApiReducer: applyReducerHash(openApiReducer, openApiInitialState),
    breadcrumbsReducer: applyReducerHash(
      breadcrumbsReducer,
      initialBreadcrumbsState
    ),
    i18nReducer: applyReducerHash(i18nReducer, i18nInitialState),
    notifications: notificationsReducer
  });
};

export default (isProd = false): Store => {
  const registry = new ReducerRegistry({}, [
    ...baseMiddlewares,
    ...prodMiddlewares,
    ...(isProd ? [] : [reduxLogger])
  ]);
  registerReducers(registry);
  return registry.getStore() as Store;
};

export const testStore = (): Store => {
  const registry = new ReducerRegistry({}, [...baseMiddlewares]);
  registerReducers(registry);
  return registry.getStore() as Store;
};
