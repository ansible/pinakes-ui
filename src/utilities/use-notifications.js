import { shallowEqual, useSelector } from 'react-redux';

const useNotification = (id) => {
  const { notificationSettings } = useSelector(
    ({ notificationSettingsReducer: { notificationSettings } }) => ({
      notificationSettings
    }),
    shallowEqual
  );

  return (
    notificationSettings &&
    notificationSettings.data &&
    notificationSettings.data.find((n) => n.id === id)
  );
};

export default useNotification;
