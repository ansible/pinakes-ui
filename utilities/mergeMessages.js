/* eslint-disable no-prototype-builtins */
/* eslint-disable no-undef */
const program = require('commander');
const fs = require('fs');
const { sync: globSync } = require('glob');
const { sync: mkdirpSync } = require('mkdirp');
const last = require('lodash/last');

let MESSAGES_PATTERN = 'build/messages/**/*.json';
let LANG_DIR = 'locales/';
let LANG_PATTERN = 'locales/**/messages.json';
let IGNORED = ['translations', 'data'];

program
  .option('-p, --pattern <value>', 'file pattern')
  .option('-I, --ignore-files <value>', 'array of ignored files')
  .option(
    '-L, --lang-pattern <value>',
    'pattern to look for files with languages'
  )
  .option('-l, --lang-dir <dir>', 'folder with languages');

const rootFolder = `${process.cwd()}/`;

program.parse(process.argv);
if (program.ignoreFiles) {
  IGNORED = program.ignoreFiles.split(',');
}

if (program.pattern) {
  MESSAGES_PATTERN = program.pattern;
}

if (program.langDir) {
  LANG_DIR = program.langDir;
}

if (program.langPattern) {
  LANG_PATTERN = `${LANG_DIR}${program.langPattern}`;
}

// Merge translated json files (es.json, fr.json, etc) into one object
// so that they can be merged with the eggregated 'en' object below
const localeMessages = (filename) => {
  const file = fs.readFileSync(filename, 'utf8');
  const messages = JSON.parse(file);
  let collection = {};
  Object.keys(messages).forEach((key) => {
    if (collection.hasOwnProperty(key)) {
      throw new Error(`Duplicate message id: ${key}`);
    }

    collection[key] = messages[key].defaultMessage;
  });
  return collection;
};

const mergedTranslations = globSync(`${rootFolder}${LANG_PATTERN}`)
  .map((filename) => {
    const locale = filename.split('/').reverse()[1];
    const file = last(filename.split('/'));
    if (!IGNORED.includes(file)) {
      return { [locale]: localeMessages(filename) };
    }
  })
  .reduce((acc, localeObj) => {
    return { ...acc, ...localeObj };
  }, {});

// Aggregates the default messages that were extracted from the example app's
// React components via the React Intl Babel plugin. An error will be thrown if
// there are messages in different components that use the same `id`. The result
// is a flat collection of `id: message` pairs for the app's default locale.
const defaultMessages = globSync(`${rootFolder}${MESSAGES_PATTERN}`)
  .map((filename) => fs.readFileSync(filename, 'utf8'))
  .map((file) => JSON.parse(file))
  .reduce((collection, descriptors) => {
    Object.keys(descriptors).forEach((key) => {
      if (collection.hasOwnProperty(key)) {
        throw new Error(`Duplicate message id: ${key}`);
      }

      collection[key] = descriptors[key].defaultMessage;
    });
    return collection;
  }, {});
// Create a new directory that we want to write the aggregate messages to
mkdirpSync(`${rootFolder}${LANG_DIR}`);

// Merge aggregated default messages with the translated json files and
// write the messages to this directory
fs.writeFileSync(
  `${rootFolder}${LANG_DIR}data.json`,
  JSON.stringify({ en: defaultMessages || {}, ...mergedTranslations }, null, 2)
);

fs.writeFileSync(
  `${rootFolder}${LANG_DIR}translations.json`,
  JSON.stringify({ ...defaultMessages } || {}, null, 2)
);
