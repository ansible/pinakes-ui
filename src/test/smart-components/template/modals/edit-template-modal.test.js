import React from 'react';
import { act } from 'react-dom/test-utils';
import thunk from 'redux-thunk';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { Modal } from '@patternfly/react-core';
import { shallowToJson } from 'enzyme-to-json';
import { MemoryRouter, Route } from 'react-router-dom';
import promiseMiddleware from 'redux-promise-middleware';
import { componentTypes } from '@data-driven-forms/react-form-renderer';
import notificationsMiddleware from '@redhat-cloud-services/frontend-components-notifications/notificationsMiddleware';
import { useIntl } from 'react-intl';
import FormRenderer from '../../../../smart-components/common/form-renderer';
import { mockApi } from '../../../../helpers/shared/__mocks__/user-login';
import EditTemplate from '../../../../smart-components/template/edit-template-modal';
import { APPROVAL_API_BASE } from '../../../../utilities/approval-constants';

describe('<EditTemplate />', () => {
  let initialProps;
  let initialState;
  const middlewares = [thunk, promiseMiddleware, notificationsMiddleware()];
  let mockStore;

  const ComponentWrapper = ({ store, children, initialEntries }) => (
    <Provider store={store}>
      <MemoryRouter initialEntries={initialEntries}>{children}</MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    localStorage.setItem('catalog_standalone', true);
    localStorage.setItem('user', 'test');
    initialProps = {
      closeUrl: 'foo',
      templateId: '123',
      objectType: 'Portfolio',
      objectName: () => 'Test Resource Name',
      pushParam: { pathname: '/foo', search: '?platform=1' }
    };
    initialState = {
      i18nReducer: {
        ...useIntl()
      },
      templateReducer: {
        templates: {
          results: [
            {
              id: '111',
              name: 'Workflow111'
            },
            {
              id: '222',
              name: 'template'
            }
          ]
        }
      }
    };
    mockStore = configureStore(middlewares);
  });

  afterEach(() => {
    mockApi.reset();
    localStorage.setItem('catalog_standalone', false);
    localStorage.removeItem('user');
  });

  it('should render correctly', async () => {
    const store = mockStore(initialState);
    mockApi
      .onGet(`${APPROVAL_API_BASE}/templates/`)
      .replyOnce(200, { results: [] });
    let wrapper;
    await act(async () => {
      wrapper = shallow(
        <ComponentWrapper store={store}>
          <EditTemplate querySelector="template" {...initialProps} />
        </ComponentWrapper>
      ).dive();
    });
    wrapper.update();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should create the edit template modal', async (done) => {
    const store = mockStore(initialState);
    mockApi.onGet(`${APPROVAL_API_BASE}/templates/123/`).replyOnce(200, {
      title: 'template',
      id: '123',
      description: 'description'
    });

    mockApi
      .onGet(`${APPROVAL_API_BASE}/notifications_settings/`)
      .replyOnce(200, {
        data: [
          {
            name: 'notification',
            id: '123',
            description: 'description'
          }
        ]
      });

    const expectedSchema = {
      fields: [
        {
          component: componentTypes.TEXT_FIELD,
          name: 'title',
          isRequired: true,
          id: 'template-title',
          label: 'Title',
          validate: [
            expect.any(Function),
            {
              message: 'Enter a name for the template',
              type: 'required'
            }
          ]
        },
        {
          component: componentTypes.TEXTAREA,
          name: 'description',
          id: 'template-description',
          label: 'Description'
        },
        {
          component: 'select',
          label: 'Process method',
          loadOptions: expect.any(Function),
          name: 'process_method',
          isClearable: true,
          isSearchable: true,
          placeholder: 'Select...'
        },
        {
          component: 'select',
          label: 'Signal method',
          loadOptions: expect.any(Function),
          isClearable: true,
          isSearchable: true,
          name: 'signal_method',
          placeholder: 'Select...'
        }
      ]
    };

    let wrapper;
    await act(async () => {
      wrapper = mount(
        <ComponentWrapper
          store={store}
          initialEntries={['/template?template=123']}
        >
          <Route
            path="/template"
            render={(args) => (
              <EditTemplate
                querySelector="template"
                {...args}
                {...initialProps}
              />
            )}
          />
        </ComponentWrapper>
      );
    });

    wrapper.update();
    const modal = wrapper.find(Modal);
    const form = wrapper.find(FormRenderer);
    expect(modal.props().title).toEqual('Edit information');
    expect(form.props().schema).toEqual(expectedSchema);
    done();
  });

  it('should submit updated template', async (done) => {
    const store = mockStore(initialState);

    mockApi
      .onGet(`${APPROVAL_API_BASE}/notifications_settings/`)
      .replyOnce(200, { data: [{ id: 'id', title: 'name' }] });

    mockApi.onGet(`${APPROVAL_API_BASE}/templates/123/`).replyOnce(200, {
      title: 'template',
      id: '123',
      description: 'description'
    });
    mockApi.onGet(`${APPROVAL_API_BASE}/templates/?title=template`).reply(200, [
      {
        name: 'template1',
        id: '111'
      },
      {
        name: 'template2',
        id: '222'
      }
    ]);
    mockApi
      .onGet(`${APPROVAL_API_BASE}/templates/?&page_size=50&page=1`)
      .reply(200, { results: [{ name: 'template1', id: '111' }] });

    mockApi.onPatch(`${APPROVAL_API_BASE}/templates/123/`).replyOnce((req) => {
      expect(JSON.parse(req.data)).toEqual({
        title: 'template',
        description: 'some-description',
        id: '123'
      });
      return [200, {}];
    });
    mockApi
      .onGet(`${APPROVAL_API_BASE}/notifications_settings/`)
      .replyOnce(200, {
        data: [
          {
            name: 'notification',
            id: '123',
            description: 'description'
          }
        ]
      });

    let onCloseMock = jest.fn().mockImplementation(() =>
      Promise.resolve({
        data: []
      })
    );

    let wrapper;
    await act(async () => {
      wrapper = mount(
        <ComponentWrapper
          store={store}
          initialEntries={['/template?template=123']}
        >
          <Route
            path="/template"
            render={(args) => (
              <EditTemplate
                {...args}
                {...initialProps}
                querySelector="template"
                postMethod={onCloseMock}
              />
            )}
          />
        </ComponentWrapper>
      );
    });
    await act(async () => {
      /**run first debounced data loading for existing templates */
      jest.runAllTimers();
      wrapper.update();
    });
    await act(async () => {
      const description = wrapper.find('textarea').first();
      description.instance().value = 'some-description';
      description.simulate('change');
    });
    wrapper.update();
    await act(async () => {
      /**run rest of paginated async request */
      jest.runAllTimers();
      wrapper.update();
    });
    await act(async () => {
      wrapper
        .find('button')
        .first()
        .simulate('click');
    });

    expect(onCloseMock).not.toHaveBeenCalled();

    await act(async () => {
      wrapper.find('form').simulate('submit');
    });

    wrapper.update();

    setImmediate(() => {
      expect(
        wrapper.find(MemoryRouter).instance().history.location.pathname
      ).toEqual('/approval/templates');
      done();
    });
  });
});
