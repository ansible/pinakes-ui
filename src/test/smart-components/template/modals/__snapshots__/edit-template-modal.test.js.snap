// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditTemplate /> should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <MemoryRouter>
    <EditTemplate
      closeUrl="foo"
      objectName={[Function]}
      objectType="Portfolio"
      pushParam={
        Object {
          "pathname": "/foo",
          "search": "?platform=1",
        }
      }
      querySelector="template"
      templateId="123"
    />
  </MemoryRouter>
</ContextProvider>
`;
